### PR TITLE
Connectivity

### PR DESCRIPTION
**Pull request by hoodmane**
I rearranged the main loop of the resolver to resolve in order by stem (using the connectivity of the algebra) as requested by Robert.

### CODE REVIEW

**Comment by hoodmane** [comment]
I disabled a bunch of asserts that were breaking things. Also, `res.extend_through_degree` is completely wrong -- I just call `res.extend_through_degree(0, 200, 0, 200);` before doing stuff because that is big enough for it not to matter that `extend_through_degree` is incorrect.

**Comment by hoodmane** [comment]
I think I significantly improved the `FreeModule` logic by using `OnceVec`'s instead of passing these tables around everywhere. The only reason I did it via passing tables before was because I wrote `FreeModule` before `OnceVec`.

**Comment by dalcde** [comment]
It seems like you messed up the commits quite a bit.

**Comment by hoodmane** [comment]
Yes. Some day maybe I'll learn how to use git.

**Comment by dalcde** [comment]
Do you mind if I force-push to fix this?

**Comment by hoodmane** [comment]
Please do.

**Comment by hoodmane** [comment]
Thanks!

File: src/resolution.rs [hunk 1]
```diff
@@ -169,18 +169,36 @@ impl<CC : ChainComplex> ResolutionInner<CC> {
         if current_differential.next_degree() > t {
             // Already computed this degree.
             return
-        } else if current_differential.next_degree() > t {
-            // Haven't computed far enough yet
-            panic!(format!("We're not ready to compute bidegree ({}, {}) yet.", s, t));
-        }
+        } 
+        // else if current_differential.next_degree() > t {
+        //     // Haven't computed far enough yet
```
Comment (dalcde) [inline src/resolution.rs]: why did you comment this out?

**Comment by dalcde** [comment]
I don't know if I messed things up but it seems to crash all the time now

File: src/resolution.rs (hunk 1 above)
Comment (hoodmane) [reply src/resolution.rs]: It was causing me trouble at some stage in debugging. I see you put back in the match on compare statement with your force push.

I would have polished this more but my thesis is due today so I had to switch to that.

**Comment by hoodmane** [comment]
The resolve_through_bidegree function is currently not correct. I have been using my own variant of that. One minute, I will try to fix it.

**Comment by dalcde** [comment]
Some questions about `FreeModule`:
 - You removed the lock on `FreeModule`; I don't think removing locks completely is the right thing to do. At some point we decided we do not impose any write locks on `OnceVec`, and instead every user of `OnceVec` must implement its own locking to prevent simultaneous writes. Currently we do not have any protection against that
 - You are using `OnceBiVec<OnceVec<...>>` for indexing `degree -> basis`. Could we not use `OnceBiVec<Vec<...>>` instead?

Otherwise I think I'm pretty in favor of the changes

**Comment by hoodmane** [comment]
I removed the lock because this is a sloppy implementation and I wasn't sure whether the lock should be on the `FreeModule` or on the `Resolution`.  If you think there should be a lock on the FreeModule, go ahead and put one back in. I haven't thought about the concurrency very carefully. This sort of sloppiness is why I submitted this as a pull request.

With the current logic, we could not use `OnceBiVec<Vec<...>>`. To do this "compute a rectangle in xy coordinates" thing, we need to make use of the fact that the kernel of ds in degree t only involves Steenrod operations on generators in degree <= t-1 and that before we add the new degree t generators, the image of ds only involves Steenrod operations on generators in degree <= t-2. In order to do this, I need to extend the tables two degrees past where we know how many generators there are and then readjust the tables in those extra degrees when we add the generators later. Thus, the need for `OnceBiVec<OnceVec<...>>`.

**Comment by dalcde** [comment]
Ah I see, you initialize the `basis_element_to_opgen` halfway before you add generators, to avoid passing the table around. I think this can potentially be problematic.

In our usage of `OnceVec`'s, there is an implicit contract that whenever we write something to a `OnceVec`, it is definite and cannot be changed. We can only push new elements. The current code violates this contract. Imagine the following situation:
 - Thread A wants to add generators to `OnceVec` in degree `t`. It calls `extend_table_entries`.
 - Thread B reads the dimension in degree `t`. This works because `basis_element_to_opgen[t]` is populated, but not in its final state. In the previous code, this would panic.
 - Thread A finally gets around to calling `add_generators`. The dimension of the module in degree `t` changes, so the dimension Thread B read is now invalid. If Thread B goes on to use this dimension, catastrophic things can happen.

In fact, the resolver code uses this behavior to do its work.

I think it is better to adopt your initial refactoring strategy where the next table entry is stored in a Mutex that `lock` obtains, which solves the other problem as well.

**Comment by hoodmane** [comment]
I think handling two different partial tables is too complicated. This current approach also works almost unchanged for algebras with stronger connectivity: if the algebra is n-connective then instead of t+1 you would use t+n in a few place.

I think it is alright to require the `Resolution` object to do sensible things. For instance, it should only have threads working on bidegrees that are independent from each other. In this case, that means that in xy coordinates it must be done with (x-1,y) and (x, y-1) before working on (x,y). I don't see that we need a lock on the `FreeModule` to enforce that: `FreeModule` can just expose this unsafe feature and the `Resolution` can wait to call `extend_step(x,y)` until it has already called `extend_step(x-1,y)` and `extend_step(x,y-1)`. This doesn't seem like a very complicated invariant to expose.

**Comment by hoodmane** [comment]
If you got the table approach, you need to add at least
```rust
FreeModule.get_dimension_with_table
FreeModule.basis_element_to_opgen_with_table
FreeModule.act_on_basis_with_table`
FreeModule.act_with_table
FreeModule.get_matrix_with_source_and_target_table
```
which I guess is sort of alright, but I got tired of passing `Option<&FreeModuleTable>`'s everywhere.

**Comment by dalcde** [comment]
I think we can simply fix the issue by checking `degree < self.num_gens.len()` in `dimension(degree)`, and add a `dimension_partial` function that allows us to access the partially computed degrees.

**Comment by dalcde** [comment]
I still think we should have locks. I very much prefer code that is safe under all possible circumstances than ones that we have to tiptoe around. I made an exception around `OnceVec` because of performance implications, but I'm still not very happy about that. Perhaps `push` should be an `unsafe` function, but this also affects things like `iter()`.

**Comment by hoodmane** [comment]
I thought about doing that, but then I would have to add:
```rust
FreeModule.dimension_partial
FreeModule.basis_element_to_opgen_partial
FreeModule.act_on_basis_partial`
FreeModule.act_partial
FreeModuleHomomorphism<M>.get_matrix_with_partial_source
FreeModuleHomomorphism<FreeModule>.get_matrix_with_partial_source_and_target
```
I don't think it would be completely unreasonable to do that but I didn't want to on my first pass.

**Comment by dalcde** [comment]
This is especially an issue now with python bindings because our users can potentially do all sorts of crazy things on their own.

**Comment by hoodmane** [comment]
If they do those things the code either crashes or produces nonsense. It will always be possible to write wrong code. I think we only need to prevent hard-to-detect wrong code. I think it is perfectly fine to document "don't do this thing it is UB" and then allow the code to misbehave quite badly if the user does it anyways (though badly in the sense of panicking or producing garbage output, not badly in the sense of filling the filesystem with garbage files or sending lots of nonsense emails). Almost every interface has some amount of this.

**Comment by dalcde** [comment]
Another possible way out is to re-introduce the `max_computed_degree` function in the `Module` trait, and make it undefined behaviour to call any functions with degree above `max_computed_degree`. It may panic (old behaviour) or produce nonsensical results.

In more user-facing python bindings we can check this before calling the rust functions, but in the Rust "core" we sidestep this for better performance.

I think I am somewhat happy with undefined behaviour that require very simple invariants.

**Comment by hoodmane** [comment]
Yes, that sounds like a good solution, I'm not really sure why we took out `max_computed_degree`: any module can opt out by just defining `max_computed_degree` to return `i32::max_value()` so there is no cost in having that as part of the interface.

**Comment by hoodmane** [comment]
Perhaps we should have a debug flag for the bindings that disables checks that are for user safety though -- when I've been debugging it's been very useful to be able to directly access internals.

**Comment by dalcde** [comment]
I've always been contemplating a `GradedObject` trait that `Algebra` and `Module` inherit. Sounds like a good time to introduce these. We can also have `BigradedObject` for chain complexes.

**Comment by hoodmane** [comment]
You can add those if you want. It's also possible to just have a bunch of different structs with the same method names. Robert has been making fun of us for writing Enterprise Quality Code lately.

**Comment by dalcde** [comment]
Yeah my main concern is that there are, like, 2 examples

**Comment by hoodmane** [comment]
Okay now it works if you just say `cargo run`.

**Comment by dalcde** [comment]
This doesn't seem to be resolving by stem...

**Comment by hoodmane** [comment]
No it isn't. I just fixed the `step_resolution` function so that it _can_ resolve by stem. I haven't changed the `resolve_through_degree` function (I am confused by the `unit_res` stuff) and so it is still resolving in the original pattern.

**Comment by hoodmane** [comment]
In particular, should I: 
(1) change `resolve_through_bidegree` to take x,y instead of s,t and resolve an xy rectangle, expect other chain complexes to behave the same or
(2) resolve an s,t rectangle but doing smaller bidegrees with smaller x coordinate first?

Option 2 is probably the fastest to implement. I can do that now.

**Comment by dalcde** [comment]
I think `resolve_through_bidegree` should resolve an `s, t` rectangle, and there should be an additional command that resolves the triangle as well; as a chain complex, the `max_computed_bidegree` doesn't see the extra triangle, and it is an explicit feature of `Resolution` that permits the user to look beyond the `max_computed_bidegree` in this case (i.e. we define the behavior, hence it is no longer undefined)

**Comment by dalcde** [comment]
fyi, I don't know how this would scale, but it now takes 40% more time to resolve S_2 up to 60

**Comment by dalcde** [comment]
I'd imagine it is easy to significantly improve this by resolving the old way until we hit the solve -1 edge

**Comment by hoodmane** [comment]
Yes that is a good idea.

**Comment by hoodmane** [comment]
I don't think there is a way to do this resolve-by-stem behavior without doing an extra copy of the matrix because the dimensions of things change between when we originally calculate the matrix and when we use the matrix. 

Using the normal approach until we hit the edge of the requested region and then filling in with this at the end seems like a way to avoid having to wait for (100, 4) before getting (75, 39) without doing these extra copies too much.

**Comment by hoodmane** [comment]
So should I rename my modified main loop and merge this by adding both main loops?
The old main loop can easily be modified to use this new way of storing kernel by bidegree if we think it's appropriate. The transition between the two approaches will be a bit tricky to get right though.

Maybe we can extract this into three functions:
```rust
get_matrix_and_row_reduce
resize_matrix
add_generators
```
So the normal main loop can look like: 
```
matrix = get_matrix_and_row_reduce(s, t)
old_kernel = get(self.kernels[s][t])
add_generators(matrix, old_kernel)
self.kernels[s+1][t] = new_kernel
if at_boundary {
    get_matrix_and_row_reduce(s, t+1)
    self.kernels[s+1][t+1] = calculated_kernel
    self.image[s][t+1] = calculated_image
}
```
and the by-stem main loop can look like:
```
old_kernel = get(self.kernels[s][t])
matrix = get(self.kernels[s][t])
resize_matrix(matrix)
add_generators(matrix, old_kernel)
get_matrix_and_row_reduce(s, t+1)
self.kernels[s+1][t+1] = calculated_kernel
self.image[s][t+1] = calculated_image
```

**Comment by dalcde** [comment]
It might be easier for the "fill in triangle" function to first resolve once along the cutoff diagonal then do the filling in? Then the usual resolve logic can remain the same.

**Comment by hoodmane** [comment]
If we extract the methods nicely, it should be fine to do it either way. I do think the main loop could use to be chopped into more manageably sized methods.

**Comment by hoodmane** [comment]
On further reflection, it is probably not the copy that is causing the performance to be so bad because I don't think the copy happens very often.
Presumably the problem is due to worse cache behavior. It's also possible that I messed up some of the arguments that prevent us from trying to row reduce empty sections of the matrix (we should be careful about those because that could do a  lot of damage to performance).

**Comment by dalcde** [comment]
Would you want to merge in the free module changes first? We can then think about how to arrange this resolution business.

One thing I would like to do is to make `ResolutionInner` a fully independent chain complex, and make `compute_through_bidegree` actually do the work instead of just asserting that we have done it. The wrapping `Resolution` struct then independently keeps track of which bidegrees it has processed.

Currently the role of `Resolution` is merely to report the results and manage the computation of products. Eventually our other interfaces will abandon `Resolution` in favor of more versatile managers, e.g. handling maps between different resolutions.

**Comment by dalcde** [comment]
And with this rearrangement, `step_resolution` will perform one of the two possible extension algorithms, depending on what we have resolved so far. We can also establish more complicated contracts about what `has_computed_bidegree` implies.

**Comment by hoodmane** [comment]
Yes I will merge the FreeModule changes in a moment. Part of the reason I haven't done that is that I'm not sure how git does partial merges but I can just do it by hand.

**Pull request merged**